neqo-http3/src/stream_type_reader.rs

```diff
@@ -139,78 +208,172 @@ mod tests {
+            &[QPACK_UNI_STREAM_TYPE_ENCODER],
+            false,
+            Ok(ReceiveOutput::NewStream(NewStreamType::Decoder)),
             true,
         );
 
-        let mut t = Test::new();
+        let mut t = Test::new(false);
         t.decode(
-            HTTP3_UNI_STREAM_TYPE_CONTROL,
-            &ReceiveOutput::NewStream(Http3StreamType::Control),
+            &[HTTP3_UNI_STREAM_TYPE_CONTROL],
+            false,
+            Ok(ReceiveOutput::NewStream(NewStreamType::Control)),
             true,
         );
 
-        let mut t = Test::new();
+        let mut t = Test::new(true);
         t.decode(
-            HTTP3_UNI_STREAM_TYPE_PUSH,
-            &ReceiveOutput::NewStream(Http3StreamType::Push),
+            &[HTTP3_UNI_STREAM_TYPE_PUSH, 0xaaaa_aaaa],
+            false,
+            Ok(ReceiveOutput::NewStream(NewStreamType::Push(0xaaaa_aaaa))),
             true,
         );
 
-        let mut t = Test::new();
+        let mut t = Test::new(false);
         t.decode(
-            0x3fff_ffff_ffff_ffff,
-            &ReceiveOutput::NewStream(Http3StreamType::Unknown),
+            &[HTTP3_UNI_STREAM_TYPE_PUSH],
+            false,
+            Err(Error::HttpStreamCreation),
+            true,
+        );
+
+        let mut t = Test::new(true);
+        t.decode(
+            &[0x3fff_ffff_ffff_ffff],
+            false,
+            Ok(ReceiveOutput::NewStream(NewStreamType::Unknown)),
             true,
         );
     }
 
     #[test]
     fn done_decoding() {
-        let mut t = Test::new();
+        let mut t = Test::new(true);
+        t.decode(
+            &[0x3fff],
+            false,
+            Ok(ReceiveOutput::NewStream(NewStreamType::Unknown)),
+            true,
+        );
+        // NewStreamTypeReader is done, it will not continue reading from the stream.
         t.decode(
-            0x3fff,
-            &ReceiveOutput::NewStream(Http3StreamType::Unknown),
+            &[QPACK_UNI_STREAM_TYPE_DECODER],
+            false,
+            Ok(ReceiveOutput::NoOutput),
             true,
         );
-        t.decode(HTTP3_UNI_STREAM_TYPE_PUSH, &ReceiveOutput::NoOutput, true);
     }
 
     #[test]
     fn decoding_truncate() {
-        let mut t = Test::new();
-        t.decode_buffer(&[0xff], &ReceiveOutput::NoOutput, false);
+        let mut t = Test::new(false);
+        t.decode_buffer(&[0xff], false, Ok(ReceiveOutput::NoOutput), false);
     }
 
     #[test]
     fn reset() {
-        let mut t = Test::new();
+        let mut t = Test::new(true);
         t.decoder.stream_reset(0x100, ResetType::Remote).unwrap();
-        t.decode(HTTP3_UNI_STREAM_TYPE_PUSH, &ReceiveOutput::NoOutput, true);
+        // after a reset NewStreamTypeReader will not read more data.
+        t.decode(
+            &[QPACK_UNI_STREAM_TYPE_DECODER],
+            false,
+            Ok(ReceiveOutput::NoOutput),
+            true,
+        );
+    }
+
+    #[test]
+    fn stream_fin() {
```

Comment:
No reason not to make this 7 different test functions.  I would only merge the two push tests.